### PR TITLE
assign mentor to mentee

### PR DESCRIPTION
**Pull request by Opio-Emmanuel-Omona**
#### How can this PR be tested
* clone the repo 
`git clone https://github.com/oneworldcoders/hm_04_26_one_world_mentors_backend.git`
* change to the. repo directory
`cd hm_04_26_one_world_mentors_backend`
* checkout this branch 
`git checkout assign_mentor`
* install necessary gems
`bundle install`
* run migrations
`rake db:migrate`
* load the seed data
`rake db:seed`
* start the server
`rails s`
* open postman hit the mentees' endpoint (authentication required)
` http://localhost:3000/mentees/1`
payload
`
{
    course_id: 1
}
`

#### Expected Result
The available and assigned mentor id should be added to the mentee.

### CODE REVIEW

File: app/controllers/assign_mentor.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class AssignMentor
+  def self.assign(mentee, course_id)
+    available_mentors = Mentor.where(available: true)
+    assigned_mentor = available_mentors.find do |mentor|
+      mentor.courses.where(id: course_id)
+    end
```
Comment (naush) [inline app/controllers/assign_mentor.rb]: This looks like a possible n+1 query problem. If you have `n` number of available mentors, then you will generate `n` queries.

File: app/controllers/mentees_controller.rb [hunk 2]
```diff
@@ -11,10 +11,10 @@ def index
   end
 
   def add_course
+    course_id = params[:course_id]
     current_mentee = Mentee.find(params[:id])
-    current_mentee.update!(course_id: params[:course_id])
+    current_mentee.update!(course_id: course_id)
+    AssignMentor.assign(current_mentee, course_id)
```
Comment (naush) [inline app/controllers/mentees_controller.rb]: I like the `AssignMentor` abstraction! Perhaps we can refactor line 14-16 into a `AssignCourse` class as well?

File: app/controllers/assign_mentor.rb (hunk 1 above)
Comment (naush) [reply app/controllers/assign_mentor.rb]: Also, is it possible to have multiple mentors teaching the same course? Do you just find the first available mentor to assign in that case?

File: spec/requests/mentees_request_spec.rb [hunk 3]
```diff
@@ -25,14 +25,27 @@
...
-        user = FactoryBot.create(:user)
-        course = FactoryBot.create(:course)
-        new_mentee = Mentee.create(id:1, user:user)
-        patch "/mentees/#{new_mentee.id}", params: { course_id: course.id }, headers: headers
+        patch "/mentees/#{@new_mentee.id}", params: { course_id: course.id }, headers: headers
         mentees = JSON.parse(response.body)
         expect(mentees["course_id"]).to eq(course.id)
       end
+
+      it 'assign mentor should add mentor to mentee table' do
```
Comment (naush) [inline spec/requests/mentees_request_spec.rb]: `Table` is an implementation detail. What is the business requirement that we are fulfilling here?

File: app/controllers/assign_mentor.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class AssignMentor
```
Comment (naush) [inline app/controllers/assign_mentor.rb]: Is there a unit test for this class?

**Comment by naush** [comment]
Consider adding those test data to seeds script so that other developers can simply run `rails db:seed` to set up their local instance! 💯

File: app/controllers/assign_mentor.rb (hunk 1 above)
Comment (Opio-Emmanuel-Omona) [reply app/controllers/assign_mentor.rb]: Yes, it is possible to have the first and therefore we pick the first available mentor for that course

File: app/controllers/assign_mentor.rb (hunk 4 above)
Comment (Opio-Emmanuel-Omona) [reply app/controllers/assign_mentor.rb]: Yes, it was in the controller tests, just moved it to it's own file

**Review by czarjulius — APPROVED** [review]

File: app/controllers/assign.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+class Assign
+  def self.mentor(mentee, course_id)
+    available_mentors = Mentor.where(available: true)
+    assigned_mentor = available_mentors.find do |mentor|
+      mentor.courses.where(id: course_id)
+    end
```
Comment (dougbradbury) [inline app/controllers/assign.rb:6]: I would suggest adding a scope to mentor for courses.  Mentor.who_teach(courses)
That way you could get to something like
available_mentors.who_teach(course)

File: app/controllers/assign.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+    mentee.update!(mentor_id: assigned_mentor.id)
```
Comment (dougbradbury) [inline app/controllers/assign.rb:8]: mentee.mentor = assigned_mentor

File: db/migrate/20200618122845_create_mentor_courses.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+class CreateMentorCourses < ActiveRecord::Migration[6.0]
+  def change
+    create_table :mentor_courses do |t|
```
Comment (dougbradbury) [inline db/migrate/20200618122845_create_mentor_courses.rb:3]: you can also use create_join_table here

File: db/migrate/20200618122845_create_mentor_courses.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+      t.references :mentor, null: false, foreign_key: true
+      t.references :course, null: false, foreign_key: true
+
+      t.timestamps
+    end
+  end
+end
```
Comment (dougbradbury) [inline db/migrate/20200618122845_create_mentor_courses.rb:10]: Do you want indices on these foreign key fields?

File: db/seeds.rb [hunk 9]
```diff
@@ -5,3 +5,72 @@
...
+
+mentees.each do |mentee|
+  mentee_ = Mentee.new(mentee)
+  Mentee.create!(mentee) if mentee_.valid?
+end
+
+mentor_courses.each do |mentor_course|
+  mentor_course = MentorCourse.new(mentor_course)
+  MentorCourse.create!(mentor_course)  if mentor_course.valid?
+end
```
Comment (dougbradbury) [inline db/seeds.rb:18]: I like the pattern we have at Campbell to use the factories to do the seeding.  Ask Kenneth about it!

**Review by kenneth051 — APPROVED** [review]

**Pull request merged**